src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,126 @@
+package htsjdk.samtools.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AsyncWriterPool implements Closeable {
+    private final ExecutorService executor;
+    private final AtomicBoolean poolClosed;
+    private final List<PooledWriter<?>> writers;
+
+    public AsyncWriterPool(final int threads) {
+        assert threads >= 1; // TODO: throw error
+
+        this.poolClosed = new AtomicBoolean(false);
+        this.writers = new ArrayList<>();
+
+        executor = Executors.newWorkStealingPool(threads);
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        for (PooledWriter<?> writer : this.writers) {
+            writer.close();
+        }
+        this.executor.shutdown();
+    }
+
+
+    // TODO: note that this is not threadsafe and multiple threads trying to call write or close should by synchronized
+    public class PooledWriter<A> implements Writer<A> {
+        private final Writer<A> writer;
+        private final BlockingQueue<A> queue;
+        private final AtomicBoolean writing;
+        private final AtomicBoolean isClosed;
+
+        private final int bufferSize;
+
+        // Holds the Future of the last task submitted to the AsyncWriterPools executor, or null if nothing has been
+        // submitted yet. Inside the Future is the number of items written and any exceptions that may have been thrown.
+        private Future<Void> currentTask;
+
+
+        public PooledWriter(final Writer<A> writer, final BlockingQueue<A> queue, final int bufferSize) {
```

Comment:
@tfenne Ah I see now what `bufferSize` is for .  Given that I would:
1. check that `bufferSize` is not larger than the queue's capacity.  If it is, then we'll never reach `bufferSize`.
2. As @tfenne suggests, get's rid of the other constructors (sorry I asked earlier) so that setting `bufferSize` is _intentional_.
3. Rephrase the docs.